### **CSCN-8000 Artificial Intelligence Algorithms and Mathematics**

#### **Lab-2**

**Submitted by : Nikhil Shankar C S**

**Id: 9026254**

#### **Practical Part A**

In [127]:
import pandas as pd
import sklearn as skl


df = pd.read_csv('data/Lab2_dataset.csv')
display(df)

df['label'].unique()

# Figuring out the meaning of Unnamed column. 
# Checking if its the word count
df['word count'] = df['text'].apply(lambda x: len(str(x).split()))
display(df)



Unnamed: 0 label                                               text  \
0            605   ham  Subject: enron methanol ; meter # : 988291\nth...   
1           2349   ham  Subject: hpl nom for january 9 , 2001\n( see a...   
2           3624   ham  Subject: neon retreat\nho ho ho , we ' re arou...   
3           4685  spam  Subject: photoshop , windows , office . cheap ...   
4           2030   ham  Subject: re : indian springs\nthis deal is to ...   
...          ...   ...                                                ...   
5166        1518   ham  Subject: put the 10 on the ft\nthe transport v...   
5167         404   ham  Subject: 3 / 4 / 2000 and following noms\nhpl ...   
5168        2933   ham  Subject: calpine daily gas nomination\n>\n>\nj...   
5169        1409   ham  Subject: industrial worksheets for august 2000...   
5170        4807  spam  Subject: important online banking alert\ndear ...   

      label_num  
0             0  
1             0  
2             0  
3             1  
4             0  
...         ...  
5166          0  
5167          0  
5168          0  
5169          0  
5170          1  

[5171 rows x 4 columns]

Unnamed: 0 label                                               text  \
0            605   ham  Subject: enron methanol ; meter # : 988291\nth...   
1           2349   ham  Subject: hpl nom for january 9 , 2001\n( see a...   
2           3624   ham  Subject: neon retreat\nho ho ho , we ' re arou...   
3           4685  spam  Subject: photoshop , windows , office . cheap ...   
4           2030   ham  Subject: re : indian springs\nthis deal is to ...   
...          ...   ...                                                ...   
5166        1518   ham  Subject: put the 10 on the ft\nthe transport v...   
5167         404   ham  Subject: 3 / 4 / 2000 and following noms\nhpl ...   
5168        2933   ham  Subject: calpine daily gas nomination\n>\n>\nj...   
5169        1409   ham  Subject: industrial worksheets for august 2000...   
5170        4807  spam  Subject: important online banking alert\ndear ...   

      label_num  word count  
0             0          67  
1             0          23  
2             0         550  
3             1          48  
4             0          70  
...         ...         ...  
5166          0         155  
5167          0         305  
5168          0          78  
5169          0         111  
5170          1         199  

[5171 rows x 5 columns]

> The unnamed column could be anything. Confirmed its not word count

In [128]:
print(f"Text : {df['text'][0]}")

Text : Subject: enron methanol ; meter # : 988291
this is a follow up to the note i gave you on monday , 4 / 3 / 00 { preliminary
flow data provided by daren } .
please override pop ' s daily volume { presently zero } to reflect daily
activity you can obtain from gas control .
this change is needed asap for economics purposes .


#### **Vectorizing Text Column**

In [129]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()

text_vector = vectorizer.fit_transform(df['text'])

print("Vector Representation:\n", text_vector.toarray())
print(f'Shape of vector : {text_vector.shape}')

print(f'Unique Word Count:\n{len(vectorizer.get_feature_names_out())}')



Vector Representation:
 [[1 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
Shape of vector : (5171, 50447)
Unique Word Count:
50447


**We can remove commonly used words that are insignificant ( Stop Words)**

In [130]:
vectorizer = CountVectorizer(stop_words='english')

text_vector = vectorizer.fit_transform(df['text'])

print("Vector Representation:\n", text_vector.toarray())
print(f'Shape of vector : {text_vector.shape}')

print(f'Unique Word Count:\n{len(vectorizer.get_feature_names_out())}')

Vector Representation:
 [[1 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
Shape of vector : (5171, 50140)
Unique Word Count:
50140


> Also we can see that "subject" is a word that is repeated in every row. So we can remove that as well.

In [131]:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

stop_words_new = list(set(ENGLISH_STOP_WORDS).union(['subject']))

vectorizer = CountVectorizer(stop_words=stop_words_new)

text_vector = vectorizer.fit_transform(df['text'])

print("Vector Representation:\n", text_vector.toarray())
print(f'Shape of vector : {text_vector.shape}')

print(f'Unique Word Count:\n{len(vectorizer.get_feature_names_out())}')

exists = 'subject' in vectorizer.get_feature_names_out()
print(exists)


Vector Representation:
 [[1 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
Shape of vector : (5171, 50139)
Unique Word Count:
50139
False


#### **Split Dataset to Training and Test**

In [132]:
# In order to split data frame into two sets 
# training and test dataset we need to import scikit learning library 'model_selection' module.
# We can just split using array split options also but then it won't be randomized. The usage of the following function from scikit 
# allows to specify a random_state parameter which promises reproducibility even when the split is randomized. ie. If same dataset is passed to the train_test_split 
# function with the same parameter values and random_state key then the split would always be identical.

from sklearn.model_selection import train_test_split

training_text_split, test_text_split, training_labels_split, test_labels_split = train_test_split(text_vector, df['label'], test_size=0.2, random_state=25)


#### **RandomForestClassifier**

In [133]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
import time
import numpy as np
import pandas as pd

def calculate_rfc_accuracy(n_estimators):
    start_time_train = time.time()
    rfc = RandomForestClassifier(n_estimators=n_estimators, random_state=42)
    rfc.fit(training_text_split, training_labels_split)
    end_time_train = time.time()
    training_time = end_time_train - start_time_train
    start_time_predict = time.time()
    predictions = rfc.predict(test_text_split)
    end_time_predict = time.time()
    prediction_time = end_time_predict - start_time_predict
    accuracy = accuracy_score(test_labels_split, predictions)
    return np.array([n_estimators, f'{training_time:.2f} seconds', f'{prediction_time:.2f} seconds', f'{accuracy * 100:.2f}%'])

rfc_analysis = np.empty((0,4))

for i in range(1,11):
    rfc_analysis = np.vstack((rfc_analysis, calculate_rfc_accuracy(i)))

rfc_analysis = np.vstack((rfc_analysis, calculate_rfc_accuracy(100)))
rfc_analysis_df = pd.DataFrame(rfc_analysis, columns=['No. of Estimators', 'Time', 'Prediction Time', 'Accuracy'])

#Commenting this to avoid hanging when running all code.
#calculate_rfc_accuracy(500)


#### **Analysis of performance of Random forest classifier**

In [134]:
display(rfc_analysis_df)

No. of Estimators           Time Prediction Time Accuracy
0                  1   0.15 seconds    0.00 seconds   91.01%
1                  2   0.28 seconds    0.00 seconds   90.72%
2                  3   0.38 seconds    0.00 seconds   93.53%
3                  4   0.61 seconds    0.00 seconds   93.72%
4                  5   0.64 seconds    0.00 seconds   95.46%
5                  6   0.78 seconds    0.00 seconds   95.56%
6                  7   0.86 seconds    0.00 seconds   96.62%
7                  8   1.02 seconds    0.01 seconds   96.43%
8                  9   1.07 seconds    0.02 seconds   96.33%
9                 10   1.24 seconds    0.01 seconds   95.94%
10               100  14.64 seconds    0.08 seconds   97.68%

> Tried running with 500 estimators and accuracy received was 97.78 compared to 97.68 with 100 estimators and time taken was around 90seconds while with 100 it was around 16~17 seconds

#### **Gaussian Naive Bayes Classifier**

In [135]:
from sklearn.naive_bayes import GaussianNB

def gaussian_nb_classifier(train_text, train_label, test_text, test_label):
    start_time = time.time()
    gnb = GaussianNB()
    gnb.fit(train_text, train_label)
    end_time = time.time()
    total_time = f'{(end_time - start_time):.2f} seconds'
    predictions = gnb.predict(test_text)
    accuracy = f'{accuracy_score(test_label, predictions) * 100}%'
    return np.array([total_time, accuracy])

gnb_analysis = np.empty((0,2))
gnb_analysis = np.vstack((gnb_analysis, gaussian_nb_classifier(training_text_split.toarray(), training_labels_split, test_text_split.toarray(), test_labels_split)))
gnb_analysis_df = pd.DataFrame(gnb_analysis, columns=['Time', 'Accuracy'])

#### **Analysis of GNB Classifier**

In [136]:
display(gnb_analysis_df)

Time           Accuracy
0  4.16 seconds  96.1352657004831%

#### **Multinomial Naive Bayes Classifier**

In [137]:
from sklearn.naive_bayes import MultinomialNB

def multinomial_nb_classifier(train_text, train_label, test_text, test_label):
    start_time = time.time()
    gnb = MultinomialNB()
    gnb.fit(train_text, train_label)
    end_time = time.time()
    total_time = f'{(end_time - start_time):.2f} seconds'
    predictions = gnb.predict(test_text)
    accuracy = f'{accuracy_score(test_label, predictions) * 100}%'
    return np.array([total_time, accuracy])

mnb_analysis = np.empty((0,2))
mnb_analysis = np.vstack((mnb_analysis, multinomial_nb_classifier(training_text_split, training_labels_split, test_text_split, test_labels_split)))
mnb_analysis_df = pd.DataFrame(mnb_analysis, columns=['Time', 'Accuracy'])

#### **Analysis of MNB Classifier**

In [138]:
display(mnb_analysis_df)

Time            Accuracy
0  0.02 seconds  98.74396135265701%

#### **Combined Analysis**

In [139]:
rfc_selected_columns = ['Time', 'Accuracy']
rfc_filter = rfc_analysis_df[rfc_analysis_df['No. of Estimators'] == '10']
rfc_final_df = rfc_filter[rfc_selected_columns]
rfc_final_df.insert(0, 'Name','RFC')
gnb_analysis_df.insert(0, 'Name','GNB')
mnb_analysis_df.insert(0, 'Name','MNB')
combined_df = pd.concat([rfc_final_df, gnb_analysis_df, mnb_analysis_df], ignore_index=True)
display(combined_df)


Name          Time            Accuracy
0  RFC  1.24 seconds              95.94%
1  GNB  4.16 seconds   96.1352657004831%
2  MNB  0.02 seconds  98.74396135265701%

#### Observations

- Multinomial Naive Bayes is lightning fast and offers highest accuracy as well. 
- RFC even though close to the accuracy of MNB the time it took was enormous and that too for a relatively small dataset.
- For real world applications the dataset will be in millions and RFC should not be the first choice given the dataset can be used with MNB. 
- GNB is also a good model since it took only 4.6 seconds to achieve 96.13% accuracy. While to achieve the same accuracy RNB had to use 100 estimators and took around 14 seconds.



#### Reasoning

- MNB calculates probability using an equation which means that once the input is given the solution can be found in linear time by sustituting values in the equation whereas RFC needs to go through the process of splitting and creating trees and assigning weights which is done recursively hence can take polynomial time complexity.